lib/ace/mode/yaml_highlight_rules.js

```diff
@@ -74,7 +74,7 @@ var YamlHighlightRules = function() {
                 token : "string", // multi line string start
                 regex : /[|>][-+\d\s]*$/,
                 onMatch: function(val, state, stack, line) {
-                    var indent = /^\s*/.exec(line)[0];
+                    var indent = /^\s*(?:[-?]\s)?/.exec(line)[0];
```

Comment:
Should this be `/^\s*(?:-\s+)?/` instead? because `?` does not start a list and the following is converted to the same json as the version with one string
```yaml
-       specialDelivery2:  >
         Follow the Yellow Brick
         Road to the Emerald City.
         Pay no attention to the
         man behind the curtain.
        blabla: This is not a string but is highlighted as one
```

Reply:
`?` is for mappings.
If changed, it should be changed to `^\s*(?:[-?]\s)?\s*`
Here's an [example](http://yaml-online-parser.appspot.com/?yaml=list1%3A%0A%20-%0A%20-%20%7C%0A%20%20block%20node%201%0A%20-%20%3E%0A%20%20block%20node%202%0A%20-%20%3E%0A%20%20%20%20%20%20block%20node%203a%0A%20%20%20%20%20%20block%20node%203b%0A%20-%20%3E%0A%20%20block%20node%204a%0A%20%20block%20node%204b%0Alist2%3A%0A-%20%20one%3A%20%3E%0A%20%20%20%20block%20node%0A%20%20%20%20block%20node%0A%20%20%20two%3A%20%3E%0A%20%20%20%20abc%0A%3F%20explicit%20key%20%23%20Empty%20value%0A%3F%20%7C%0A%20%20block%20key%201%0A%3A%20-%20one%20%23%20Explicit%20compact%0A%20%20-%20two%20%23%20block%20value%0A%3F%20%7C%0A%20%20block%20key%202%0A%3A%20-%20%3E%20%0A%20%20%20one%0A%20%20-%20two%20%23%20block%20value%0A&type=json), I think this PR is an improvement but it doesn't cover all cases, for example these snippets from the example are not correctly highlighted:
```yaml
list1:
 - >
  block node 4a
  block node 4b
```
```yaml
? |
  block key 2
: - > 
   one
  - two # block value
```

Reply:
inteeresting, so to correctly highlight something like 
```yaml
?        |
 block key #1
 kkk
: - |
   one 
  - |
   tw #o 
  - ? | 
     as #d
  - 
     x: xx
     r: xx
     z: sss
     zdd: dddd
     "block key 2" : 
       - two # block value
       - ? | 
          as #d
         : |
          asdadas #d
          asd
         a: 2
       - ? | 
          asdas #d
         :  |
          xx

```

we need to do 

```js
                    var indent = /^\s*((:\s*)?-(\s*\?)?)?/.exec(line)[0]
                        .replace(/\S\s*$/, "");
```
?

Reply:
Well, now this broke:
```yaml
list2:
-  one: >
    block node
    block node
   two: >
    abc
```

Maybe we need to calculate the minimum allowed indentation level when a multiline string and also check indentation of the first row in the multi-line string.
Here's an updated [example](http://yaml-online-parser.appspot.com/?yaml=list1%3A%0A+-+%7C%0A++block+node+1+(minimum+allowed+indentation)%0A+-+%3E+%23empty%0Alist2%3A%0A-++one%3A+%3E%0A++++block+node+(minimum+allowed+indentation)%0A++++++block+node+(additionalindentation+on+2nd+row+adds+spaces)%0A+++two%3A+%3E%0A++++abc%0A%3F+explicit+key+%23+Empty+value%0A%3F+%7C%0A+block+key+1%0A%3A+-+one+%23+Explicit+compact%0A++-+two+%23+block+value%0A%3F+%7C%0A+block+key+2%0A%3A+-+%3E+%0A+++one%0A++-+two+%23+block+value%0A&type=json).

Reply:
to fix this we can use:
```js
var indent = /^\s*((:\s*)?-(\s*[^|>])?)?/.exec(line)[0]
                        .replace(/\S\s*$/, "");
```

updating indentation on first line is more complex and requires adding a new state

```js
         {
                token : "string", // multi line string start
                regex : /[|>][-+\d\s]*$/,
                onMatch: function(val, state, stack, line) {
                    var indent = /^\s*((:\s*)?-(\s*[^|>])?)?/.exec(line)[0]
                        .replace(/\S\s*$/, "");debugger
                    var nextIndent = parseInt(/\d+[\s+-]*$/.exec(line))
                    var next = this.next;
                    
                    if (nextIndent)
                        indent += " ".repeat(nextIndent - 1)
                    else
                        next += "Pre"
                    if (stack.length < 1) {
                        stack.push(next);
                        stack.push(indent.length);
                    } else {
                        stack[0] = next;
                        stack[1] = indent.length;
                    }
                    return this.token;
                },
                next : "mlString"
            }
```
```js
        "mlStringPre" : [
            {
                token : "indent",
                regex : /^\s*$/
            }, {
                token : "indent",
                regex : /^\s*/,
                onMatch: function(val, state, stack) {
                    var curIndent = stack[1];

                    if (curIndent >= val.length) {
                        this.next = "start";
                        stack.splice(0);
                    }
                    else {
                        stack[1] = val.length - 1;
                        this.next = "mlString";
                    }
                    return this.token;
                },
                next : "mlString"
            }, {
                defaultToken : "string"
            }
        ]
```
so i am not sure if it is worth that or not.

Reply:
I'd say it's worth giving it a try at least, but note that indentation should be deduced only by counting spaces, not by counting any whitespace, so we should match `/^ */`

However, it get's worse / morecomplicated, https://yaml.org/spec/1.2/spec.html#id2793979:
 - Indentation should be deduced from the fist _non-empty_ line, not from the next line
 - Indentation can be manually specified

Reply:
Thanks for the detailed explanation, 32e531d seems to handle all of these cases. Please let me know if you see anything that is still broken with it